Rewrite the programming guide to use the new guestbook (#168)
dbos-inc/dbos-docs: docs/getting-started/quickstart-programming.md

title: Programming Guide
---

> Before starting this tutorial, we recommend finishing the [quickstart](./quickstart.md). We'll modify the quickstart example to illustrate the core building blocks of [DBOS Transact](https://github.com/dbos-inc/dbos-ts). In the end, we'll create an app that reliably records events across two different systems: the Postgres database and a third-party API. The app will write to both systems consistently, even if we crash it between writes.

Reviewer: Maybe also mention "We'll walk you through how to set up the app and its database to run it locally"

Author: Rewritten; take a look

> ## Serving Endpoints

Reviewer: ```suggestion
## Serving HTTP Requests
```

Let's start with a simple HTTP GET handler to greet friends. In your app folder, change the file `src/operations.ts` to contain only the following:

```javascript
import { HandlerContext, GetApi } from '@dbos-inc/dbos-sdk';

export class Greetings {

>   @GetApi('/greeting/:friend')

Author: Do you mean in addition to /greeting or instead of?

>   @GetApi('/greeting/:friend')

Reviewer: In addition to `/greeting`

>   @GetApi('/greeting/:friend')

Reviewer 2: I really think we should remove it. While it's nice to have, It crowds the code examples a lot and makes them hard to read.

>   @GetApi('/greeting/:friend')

Reviewer 2: The first ode example is more than half readme code that's not directly relevant to anything in the test. Makes it hard to figure out what's going on.

title: Programming Guide
---

> This tutorial assumes you have finished the [Quickstart](./quickstart.md) and you have a [DBOS Transact](https://github.com/dbos-inc/dbos-ts) app running locally. In this guide we'll modify that example to reliably record events across two different systems: Postgres and a third-party API. This app will write to both systems consistently, even if interrupted and restarted at any point.

Reviewer 2: ```suggestion
This tutorial assumes you have finished the [Quickstart](./quickstart.md) and you have a [DBOS Transact](https://github.com/dbos-inc/dbos-ts) app running locally. In this guide we'll modify that example to reliably record events across two different systems: Postgres and a third-party API. This app will write to both systems consistently, even if interrupted or restarted at any point.
```

> ## Serving HTTP Requests

Reviewer 2: I wonder if we should number these steps as we did in the quickstart? It seemed effective there.

Author: done

[info]: DBOS Admin Server is running at http://localhost:3001
```

> To see that your application is working, visit this URL in your browser: [http://localhost:3000/greeting/Mike](http://localhost:3000/greeting/Mike). You should see the message `Greetings, Mike!`. If you replace Mike with a different name, your application will greet that name instead. To learn more about HTTP serving in DBOS, see [HTTP Serving Tutorial](../tutorials/http-serving-tutorial).

Reviewer 2: ```suggestion
To see that your application is working, visit this URL in your browser: [http://localhost:3000/greeting/Mike](http://localhost:3000/greeting/Mike). You should see the message `Greetings, Mike!`. If you replace Mike with a different name, your application will greet that name instead. To learn more about HTTP serving in DBOS, see our [HTTP Serving Tutorial](../tutorials/http-serving-tutorial).
```

  }

  @GetApi('/greeting/:friend')
  static async Greeting(ctxt: HandlerContext, friend: string) {
    const noteContent = `Thank you for being awesome, ${friend}!`;
    await ctxt.invoke(Greetings).InsertGreeting(
      { name: friend, note: noteContent }
    );
    return noteContent;
  }
}
```

Here we define a `GreetingRecord` interface to match a row of data in our `greetings` table. We then define a `@Transaction` called `InsertGreeting` that puts a new `GreetingRecord` into the `greetings` table. Finally, we add a line to the GET API function `Greeting` to invoke `InsertGreeting` with the provided `name` and a welcoming `note`.

Stop your app with CTRL+C. Rebuild with `npm run build` and start with `npx dbos start`. Make a few visits to the greeting URL in your browser, i.e. http://localhost:3000/greeting/Mike. With every new visit, the app should print this to the console:
```
[info]: Greeting to Mike recorded in the database! 
```

You can then query your Postgres database to see the greeting records. It may look something like this:
```
psql -h localhost -U postgres -d hello -c "select * from greetings"; 
```

> The database name `hello` above will vary with your application name. Your `host`, `port` or `username` may also be differnt depending on how you configured Postgres in the Quickstart.

Reviewer: I think we want to drop this block. It breaks the flow of this programming guide, and we want to really focus on programming here.

Author: ...we could... but when I was personally going thru the guide. I ran the postgres thru docker. And I wasn't sure how exactly it was set up... and it took me like 30 minutes to discover that I can do this -and I found it very useful....

Reviewer: That's fair, but it may belong to the quickstart guide where we introduce Postgres setup.

Author: Yep, moved this section to quickstart in the latest commit; have a look

:::

## Interacting with External Services

Let's say we want to use a third-party service to send our greeting, or record it in a remote system. For this example, you can use our demo Guestbook app. It lets you generate a key and use it to record a few temporary greetings.

### Create a Guestbook Key
To create a new key, visit https://demo-guestbook.cloud.dbos.dev/key It should output a 36-character sequence like so `12345abc-1234-5678-1234-567890abcdef`. Yours will be different. You can pass this key to your app as a config variable. In your app folder, edit the file `dbos-config.yaml`. Add a new `env:` section at the bottom with the variable `GUESTBOOK_KEY` set to your key in quotes:
```yaml
env:
  GUESTBOOK_KEY: 'your-key-value-here'
```

For example, if your key value is `12345abc-1234-5678-1234-567890abcdef` then you would add:
```yaml
env:
  GUESTBOOK_KEY: '12345abc-1234-5678-1234-567890abcdef'
```
::::tip

> In production, to avoid storing keys in cleartext and to make rotation easier you could change this entry to `GUESTBOOK_KEY: ${ENV_GUESTBOOK_KEY}`. You would then set `ENV_GUESTBOOK_KEY` in your enviornment prior to running `npx dbos start`. This also applies to cloud deployment: the variable is expanded at `npx dbos-cloud app deploy` time and securely passed to the cloud.

Reviewer: Provide a link to `/api-reference/configuration#environment-variables` so we can greatly shorten this paragraph.

Author: done. I added a couple words to the configuration section to explain this a little more

```javascript
import { TransactionContext, Transaction, HandlerContext, GetApi } from '@dbos-inc/dbos-sdk';
import { Knex } from 'knex';

interface GreetingRecord {
  name: string;
  note: string;
}

export class Greetings {
  @GetApi('/') // Serve a quick readme for the app
  static async readme(_ctxt: HandlerContext) {
    const readme = '<html><body><p>' +
      'Welcome! Visit the route /greeting/:name to be greeted!<br>' +
      'For example, visit <a href="/greeting/dbos">/greeting/dbos</a>.<br>' +
      '</p></body></html>';
    return Promise.resolve(readme);
  }

  @Transaction()
  static async InsertGreeting(ctxt: TransactionContext<Knex>, gr: GreetingRecord) {
    await ctxt.client('greetings').insert(gr);
    ctxt.logger.info(`Greeting to ${gr.name} recorded in the database!`);
  }

  @GetApi('/greeting/:friend')
  static async Greeting(ctxt: HandlerContext, friend: string) {
    const noteContent = `Thank you for being awesome, ${friend}!`;
    await ctxt.invoke(Greetings).InsertGreeting(
      { name: friend, note: noteContent }
    );
    return noteContent;
  }
}
```

> Here we define a `GreetingRecord` interface to match a row of data in our `greetings` table. We then define a `@Transaction` called `InsertGreeting` that puts a new `GreetingRecord` into `greetings`. Finally, we add a line to the GET API function `Greeting` to invoke `InsertGreeting` with the provided `name` and a welcoming `note`.

Reviewer 2: I'd add another sentence after the second sentence that mentions that we insert the new record using Knex, but also support Prisma, TypeORM, and raw SQL. Experienced devs will immediately wonder "what's this weird interface."

Reviewer 2: I see we already have that note further down--it's good, but move it upwards

Author: done

Stop your app with CTRL+C. Rebuild with `npm run build` and start with `npx dbos start`. Make a few visits to the greeting URL in your browser, i.e. http://localhost:3000/greeting/Mike. With every new visit, the app should print this to the console:
```
[info]: Greeting to Mike recorded in the database! 
```

> ## Reading from the Database

Reviewer 2: I'd remove this section. All it introduces is `readOnly`, which is a pure optimization that's too advanced for the programming guide in my opinion.

Author: Leaving in as discussed offline

You can add another GET API function to read all the greetings from the database like so:

```javascript
//export class Greetings {
//...
  @Transaction({readOnly: true})
  @GetApi('/greetings')
  static async allGreetings(ctxt: TransactionContext<Knex>) {
    return await ctxt.client('greetings').select('*') as GreetingRecord[];
  }
//}
```

Here we use `@GetApi` and `@Transaction` together. This transaction only reads data so we mark it as `{readOnly: true}`. This enables DBOS to execute it faster, with fewer database round-trips.

:::info

> In this quickstart, we run queries using [Knex query builder](https://knexjs.org/guide/query-builder.html). DBOS Transact also supports [Knex.raw](https://knexjs.org/guide/raw.html), [TypeORM](https://typeorm.io/), and [Prisma](https://www.prisma.io/docs/orm/prisma-client).

Reviewer 2: Oh, I see, we already have this note--put it further up then.

> In this quickstart, we run queries using [Knex query builder](https://knexjs.org/guide/query-builder.html). DBOS Transact also supports [Knex.raw](https://knexjs.org/guide/raw.html), [TypeORM](https://typeorm.io/), and [Prisma](https://www.prisma.io/docs/orm/prisma-client).

Reviewer 2: ```suggestion
In this quickstart, we run queries using the [Knex query builder](https://knexjs.org/guide/query-builder.html). DBOS Transact also supports [Knex.raw](https://knexjs.org/guide/raw.html), [TypeORM](https://typeorm.io/), and [Prisma](https://www.prisma.io/docs/orm/prisma-client).
```

Reviewer 2: Change the links from their websites to our tutorials?

:::

## Interacting with External Services

> Now suppose we also want to record our greetings in a remote system. This could be something like sending an email or writing to some storage API. For this example, we'll use a demo DBOS Guestbook app. It lets us generate an API key and use it to record temporary greetings.

Reviewer 2: ```suggestion
Now, suppose we also want to send our greetings to a remote system. In this example, we'll use a demo DBOS Guestbook app. It lets us generate an API key and use it to record temporary greetings.
```

### Create a Guestbook Key

> To create a new key, visit https://demo-guestbook.cloud.dbos.dev/key It should output a 36-character sequence like so `12345abc-1234-5678-1234-567890abcdef`. Yours will be different. You can pass this key to your app as a config variable. In your app folder, edit the file `dbos-config.yaml`. Add a new `env:` section at the bottom with the variable `GUESTBOOK_KEY` set to your key in quotes:

Reviewer 2: ```suggestion
To create a new key, visit https://demo-guestbook.cloud.dbos.dev/key. It should output a 36-character sequence like `12345abc-1234-5678-1234-567890abcdef` (yours will be different).

You can pass this key to your app as a config variable. In your app folder, edit the file `dbos-config.yaml`. Add a new `env:` section at the bottom with the variable `GUESTBOOK_KEY` set to your key in quotes:
```

> ```yaml

Reviewer 2: This is clever!

env:
  GUESTBOOK_KEY: 'your-key-value-here'
```

> For example, if your key value is `12345abc-1234-5678-1234-567890abcdef` then you would add:

Reviewer 2: ```suggestion
For example, if your key value is `12345abc-1234-5678-1234-567890abcdef` then you should add:
```

```yaml
env:
  GUESTBOOK_KEY: '12345abc-1234-5678-1234-567890abcdef'
```
::::tip

> In production, to avoid storing keys in cleartext and to make rotation easier you could change this to `GUESTBOOK_KEY: ${ENV_GUESTBOOK_KEY}`. You would then set `ENV_GUESTBOOK_KEY` in your enviornment prior to starting or deploying the app. See the [Configuration Guide](../api-reference/configuration) for more details.

Reviewer 2: ```suggestion
In production, we recommend storing the key in an environment variable instead of plaintext. To do this, change the configuration to `GUESTBOOK_KEY: ${ENV_GUESTBOOK_KEY}` and set `ENV_GUESTBOOK_KEY` in your environment prior to starting or deploying the app. See the [Configuration Guide](../api-reference/configuration#environment-variables) for more details.
```

::::

### Sign the Guestbook from the App

> In DBOS, we strongly recommend that all calls to third-party APIs are wrapped in [Communicators](../tutorials/communicator-tutorial). Let's add one to our app to POST each greeting to the Guestbook using the `GUESTBOOK_KEY` config. Change your `src/operations.ts` to contain the following:

Reviewer 2: ```suggestion
In DBOS, we strongly recommend wrapping all calls to third-party APIs in [Communicators](../tutorials/communicator-tutorial). Let's add one to our app to POST each greeting to the Guestbook using the `GUESTBOOK_KEY` config. Change your `src/operations.ts` to contain the following:
```

  //Omitted for brevity: @GetApi('/greetings') //read greetings from database
  
  @Communicator()
  static async SignGuestbook(ctxt: CommunicatorContext, name: string) {
    const url = 'https://demo-guestbook.cloud.dbos.dev/record_greeting';
    const payload = {
      'key':  process.env.GUESTBOOK_KEY, //set in dbos-config.yaml
      'name': name
    };
    const headers = { 'Content-Type': 'application/json' };
    const response = await axios.post(url, payload, { headers: headers });
    ctxt.logger.info(`Signed the Guestbook: ${JSON.stringify(response.data)}`);
  }

  @Transaction()
  static async InsertGreeting(ctxt: TransactionContext<Knex>, gr: GreetingRecord) {
    await ctxt.client('greetings').insert(gr);
    ctxt.logger.info(`Greeting to ${gr.name} recorded in the database!`);
  }

  @GetApi('/greeting/:friend')
  static async Greeting(ctxt: HandlerContext, friend: string) {
    const noteContent = `Thank you for being awesome, ${friend}!`;
    await ctxt.invoke(Greetings).SignGuestbook(friend);
    await ctxt.invoke(Greetings).InsertGreeting(
      { name: friend, note: noteContent }
    );
    return noteContent;
  }
}
```

> Note the new `@Communicator` function called `SignGuestbook` that uses `axios` to send the request. If it does not receive a successful response, it will automatically retry up to 3 times with exponential backoff. This is configurable [here](https://docs.dbos.dev/tutorials/communicator-tutorial#configurable-retries). Stop your app with CTRL+C, rebuild with `npm run build` and start your application with `npx dbos start`. Make a few visits to the greeting URL in your browser, i.e. http://localhost:3000/greeting/Mike. With every new visit, the app should now print a line for the guestbook signature along with the database record:

Reviewer 2: Is axios installed by default?  Could we use `fetch` instead?

Author: changed to fetch

> Note the new `@Communicator` function called `SignGuestbook` that uses `axios` to send the request. If it does not receive a successful response, it will automatically retry up to 3 times with exponential backoff. This is configurable [here](https://docs.dbos.dev/tutorials/communicator-tutorial#configurable-retries). Stop your app with CTRL+C, rebuild with `npm run build` and start your application with `npx dbos start`. Make a few visits to the greeting URL in your browser, i.e. http://localhost:3000/greeting/Mike. With every new visit, the app should now print a line for the guestbook signature along with the database record:

Reviewer 2: ```suggestion
Note the new `@Communicator` function called `SignGuestbook` that uses `axios` to send the request. If it does not receive a successful response, it will automatically retry up to 3 times with exponential backoff. This is configurable [here](https://docs.dbos.dev/tutorials/communicator-tutorial#configurable-retries). Stop your app with CTRL+C, rebuild with `npm run build` and start your application with `npx dbos start`. Make a few visits to the greeting URL in your browser, i.e. http://localhost:3000/greeting/Mike. With every new visit, the app should now print first that it's recorded the greeting in the guestbook, then that it's recorded the greeting in the database.
```

  
  @Communicator()
  static async SignGuestbook(ctxt: CommunicatorContext, name: string) {
    const response = await fetch('https://demo-guestbook.cloud.dbos.dev/record_greeting', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json'},
      body: JSON.stringify({ 'key': process.env.GUESTBOOK_KEY, 'name': name})
    });
    const responseStr = JSON.stringify(await response.json());
    if (!response.ok) {
      throw new DBOSResponseError(responseStr);
    }
    ctxt.logger.info(`>>> STEP 1: Signed the Guestbook: ${responseStr}`);
  }

  @Transaction()
  static async InsertGreeting(ctxt: TransactionContext<Knex>, gr: GreetingRecord) {
    await ctxt.client('greetings').insert(gr);
    ctxt.logger.info(`>>> STEP 2: Greeting to ${gr.name} recorded in the database!`);
  }

  @GetApi('/greeting/:friend')
  static async Greeting(ctxt: HandlerContext, friend: string) {
    const noteContent = `Thank you for being awesome, ${friend}!`;
    await ctxt.invoke(Greetings).SignGuestbook(friend);
    await ctxt.invoke(Greetings).InsertGreeting(
      { name: friend, note: noteContent }
    );
    return noteContent;
  }
}
```

> We add a new `@Communicator` function called `SignGuestbook` that uses `fetch` to send an HTTP POST request to the guestbook to record a greeting. If this communicator throws an error, it is automatically retried up to 3 times with exponential backoff. This is configurable [here](https://docs.dbos.dev/tutorials/communicator-tutorial#configurable-retries).

Reviewer 2: ```suggestion
We add a new `@Communicator` function called `SignGuestbook` that uses `fetch` to send an HTTP POST request to the guestbook to record a greeting. If the communicator throws an error, it is automatically retried up to 3 times with exponential backoff. This is configurable [here](https://docs.dbos.dev/tutorials/communicator-tutorial#configurable-retries).
```

```javascript
import { TransactionContext, Transaction, HandlerContext, GetApi } from '@dbos-inc/dbos-sdk';
import { Knex } from 'knex';

interface GreetingRecord {
  name: string;
  note: string;
}

export class Greetings {
  @GetApi('/') // Serve a quick readme for the app
  static async readme(_ctxt: HandlerContext) {
    const readme = '<html><body><p>' +
      'Welcome! Visit the route /greeting/:name to be greeted!<br>' +
      'For example, visit <a href="/greeting/dbos">/greeting/dbos</a>.<br>' +
      '</p></body></html>';
    return Promise.resolve(readme);
  }

  @Transaction()
  static async InsertGreeting(ctxt: TransactionContext<Knex>, gr: GreetingRecord) {
    await ctxt.client('greetings').insert(gr);
    ctxt.logger.info(`Greeting to ${gr.name} recorded in the database!`);
  }

  @GetApi('/greeting/:friend')
  static async Greeting(ctxt: HandlerContext, friend: string) {
    const noteContent = `Thank you for being awesome, ${friend}!`;
    await ctxt.invoke(Greetings).InsertGreeting(
      { name: friend, note: noteContent }
    );
    return noteContent;
  }
}
```

> Here we define a `GreetingRecord` interface to match a row of data in our `greetings` table. We then define a `@Transaction` called `InsertGreeting` that puts a new `GreetingRecord` into `greetings`. Finally, we add a line to the GET API function `Greeting` to invoke `InsertGreeting` with the provided `name` and a welcoming `note`.

Reviewer 2: ```suggestion
Here we define a `GreetingRecord` interface matching a row of data in our `greetings` table. We then define a `@Transaction` called `InsertGreeting` that inserts a new `GreetingRecord` into `greetings`. Finally, we add a line to the GET API function `Greeting` to invoke `InsertGreeting` with the provided `name` and a welcoming `note`.
```

### 4.1. Create a Guestbook Key
To generate a guestbook API key, visit https://demo-guestbook.cloud.dbos.dev/key. It should output a 36-character sequence like `12345abc-1234-5678-1234-567890abcdef` (yours will be different).

You can pass this key to your app as a config variable. In your app folder, edit the file `dbos-config.yaml`. Add a new `env:` section at the bottom with the variable `GUESTBOOK_KEY` set to your key in quotes:
```yaml
env:
  GUESTBOOK_KEY: 'your-key-value-here'
```

For example, if your key is `12345abc-1234-5678-1234-567890abcdef` then you should add:
```yaml
env:
  GUESTBOOK_KEY: '12345abc-1234-5678-1234-567890abcdef'
```
::::tip
In production, we recommend storing API keys and other secrets in an environment variable instead of plaintext. To do this, change the configuration to `GUESTBOOK_KEY: ${ENV_GUESTBOOK_KEY}` and set `ENV_GUESTBOOK_KEY` in your environment prior to starting or deploying the app. See the [Configuration Guide](../api-reference/configuration#environment-variables) for more details.
::::

### 4.2. Sign the Guestbook from the App

> In DBOS, we strongly recommend wrapping all calls to third-party APIs in [Communicators](../tutorials/communicator-tutorial). Let's add a communicator to our app to POST each greeting to the Guestbook using the `GUESTBOOK_KEY` config. Change your `src/operations.ts` to contain the following:

Reviewer 2: ```suggestion
Let's update our app to POST each greeting to the guestbook.
In DBOS, we strongly recommend wrapping all such calls to third-party APIs in [Communicators](../tutorials/communicator-tutorial).
Change your `src/operations.ts` to contain the following:
```

  
  @Communicator()
  static async SignGuestbook(ctxt: CommunicatorContext, name: string) {
    const response = await fetch('https://demo-guestbook.cloud.dbos.dev/record_greeting', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json'},
      body: JSON.stringify({ 'key': process.env.GUESTBOOK_KEY, 'name': name})
    });
    const responseStr = JSON.stringify(await response.json());
    if (!response.ok) {
      throw new DBOSResponseError(responseStr);
    }
    ctxt.logger.info(`>>> STEP 1: Signed the Guestbook: ${responseStr}`);
  }

  @Transaction()
  static async InsertGreeting(ctxt: TransactionContext<Knex>, gr: GreetingRecord) {
    await ctxt.client('greetings').insert(gr);
    ctxt.logger.info(`>>> STEP 2: Greeting to ${gr.name} recorded in the database!`);
  }

  @GetApi('/greeting/:friend')
  static async Greeting(ctxt: HandlerContext, friend: string) {
    const noteContent = `Thank you for being awesome, ${friend}!`;
    await ctxt.invoke(Greetings).SignGuestbook(friend);
    await ctxt.invoke(Greetings).InsertGreeting(
      { name: friend, note: noteContent }
    );
    return noteContent;
  }
}
```

> Note the new `@Communicator` function called `SignGuestbook` that uses a `fetch` call to send the request. If this communicator throws an error, it will be automatically retried up to 3 times with exponential backoff. This is configurable [here](https://docs.dbos.dev/tutorials/communicator-tutorial#configurable-retries). Stop your app with CTRL+C, rebuild with `npm run build` and start your application with `npx dbos start`. Make a few visits to the greeting URL in your browser, i.e. http://localhost:3000/greeting/Mike. With every new visit, the app should now print first that it's recorded the greeting in the guestbook, then that it's recorded the greeting in the database.

Reviewer 2: ```suggestion
We add a new `@Communicator` function called `SignGuestbook` that uses a `fetch` to send an HTTP POST request recording each greeting in the guestbook. If this communicator throws an error, it will be automatically retried up to 3 times with exponential backoff. This is configurable [here](https://docs.dbos.dev/tutorials/communicator-tutorial#configurable-retries). Stop your app with CTRL+C, rebuild with `npm run build` and start your application with `npx dbos start`. Make a few visits to the greeting URL in your browser, i.e. http://localhost:3000/greeting/Mike. With every new visit, the app should now print first that it's recorded the greeting in the guestbook, then that it's recorded the greeting in the database.
```

```
[info]: >>> STEP 1: Signed the Guestbook: {"ip_address":"...","greeted_name":"Mike","greeted_ts":"..."}
[info]: >>> STEP 2: Greeting to Mike recorded in the database!
```

You can visit the URL `https://demo-guestbook.cloud.dbos.dev/greetings/your-key-value` to see all the Guestbook greetings made with your key. Old greetings and keys are removed after a few days.

## 5. Composing Reliable Workflows

> Next, we want to make our app **reliable**: guarantee that it inserts exactly one database record per guestbook signature, even if stopped and restarted. DBOS makes this easy with [workflows](../tutorials/workflow-tutorial.md). To see them in action, change your `src/operations.ts` like so:

Reviewer 2: ```suggestion
Next, we want to make our app **reliable**: guarantee that it inserts exactly one database record per guestbook signature, even if interrupted or restarted. DBOS makes this easy with [workflows](../tutorials/workflow-tutorial.md). To see them in action, change your `src/operations.ts` like so:
```

[info]: Greeting to Mike recorded in the database!
```

> If this were not a DBOS `@Workflow` you would expect the app to restart with a "clean slate." There would be one record in the guestbook without matching records in Postgres. But this workflow automatically resumes where it left off and finishes recording the greeting in the database. This reliability is a core feature of DBOS: workflows always continue execution from the last completed step and run completion. To learn more about workflows, check out our [tutorial](../tutorials/workflow-tutorial.md) and [explainer](../explanations/how-workflows-work.md).

Reviewer 2: ```suggestion
If this were not a DBOS `@Workflow` you would expect the app to restart with a "clean slate," leaving behind one record in the guestbook without a matching record in Postgres. However, this workflow automatically resumes from where it left off and finishes recording the greeting in the database. This reliability is a core feature of DBOS: workflows always run to completion, executing each of their steps exactly-once. To learn more about workflows, check out our [tutorial](../tutorials/workflow-tutorial.md) and [explainer](../explanations/how-workflows-work.md).
```

Author: in discussion on some of the verbiage

> Also, note that we start the workflow asynchronously using `startWorkflow`. This returns the response to the caller as soon as the workflow starts, without waiting for the workflow to return. DBOS guarantees that the workflow continues to process to completion. This behavior is preferred when the caller expects a fast response, such as with a [payment webhook](https://www.dbos.dev/blog/open-source-typescript-stripe-processing). To make it synchronous, you can switch `startWorkflow` with `invokeWorkflow`.

Reviewer 2: `startWorkflow` is important, but too advanced for this tutorial I think. I'd recommend changing back to `invokeWorkflow` and removing this paragraph.

Author: Leaving in, as discussed

Reviewer 2: I'd still make it an infobox to stress that it's optional advanced information

Author: done

> Note that we start the workflow asynchronously using `startWorkflow` in the `GreetingWorkflow` invocation. This returns the response to the caller before the workflow has a chance to finish. The caller can then check on workflow completion after the fact, by calling `/greetings` for example. This behavior is preferred when the caller expects a fast response, such as with a [payment webhook](https://www.dbos.dev/blog/open-source-typescript-stripe-processing). You can change this switching `startWorkflow` to `invokeWorkflow`. You can then edit `GreetingWorkflow` to return a value that `Greeting` passes back to the client.

Reviewer: This paragraph can be clearer, emphasizing we return after we've marked the workflow has been started.
```suggestion
Note that we start the workflow asynchronously using `startWorkflow` in the `GreetingWorkflow` invocation. This returns the response to the caller as soon as the workflow starts, without waiting for the workflow to return. DBOS guarantees that the workflow continues to process to completion asynchronously. This behavior is preferred when the caller expects a fast response, such as with a [payment webhook](https://www.dbos.dev/blog/open-source-typescript-stripe-processing). You can change this switching `startWorkflow` to `invokeWorkflow`. You can then edit `GreetingWorkflow` to return a value that `Greeting` passes back to the client.
```

[info]: Press Control + C to interrupt the workflow...
[info]: Press Control + C to interrupt the workflow...
[info]: Press Control + C to interrupt the workflow...
[info]: >>> STEP 2: Greeting to Mike recorded in the database!
```

> If this were not a DBOS `@Workflow` you would expect the app to restart with a "clean slate." You would see a `STEP 1` and no `STEP 2`. But this workflow automatically resumes where it left off and finishes recording the greeting in the database. This reliability is a core feature of DBOS: workflows always continue execution from the last completed step and run to completion. To learn more about workflows, check out our [tutorial](../tutorials/workflow-tutorial.md) and [explainer](../explanations/how-workflows-work.md).

Reviewer 2: ```suggestion
If this were not a DBOS `@Workflow` you would expect the app to restart with a "clean slate" and completely forget about your interrupted workflow. However, DBOS automatically resumes your workflow from where it left off and records in the database the greeting you earlier recorded in the guestbook. This reliability is a core feature of DBOS: workflows always continue execution from the last completed step and run to completion. To learn more about workflows, check out our [tutorial](../tutorials/workflow-tutorial.md) and [explainer](../explanations/how-workflows-work.md).
```
